pkg/backend/path_creds.go

```diff
@@ -68,23 +68,28 @@ func (b *backend) credsUpdateOperation(ctx context.Context, req *logical.Request
 
 	key := credKey(data.Get("name").(string))
 
-	code, ok := data.GetOk("code")
-	if !ok {
-		return logical.ErrorResponse("missing code"), nil
+	var tok *oauth2.Token
+
+	if code, ok := data.GetOk("code"); ok {
+		cb := p.NewExchangeConfigBuilder(c.ClientID, c.ClientSecret)
+
+		if redirectURL, ok := data.GetOk("redirect_url"); ok {
+			cb = cb.WithRedirectURL(redirectURL.(string))
+		}
+
+		tok, err = cb.Build().Exchange(ctx, code.(string))
+		if _, ok := err.(*oauth2.RetrieveError); ok {
+			return logical.ErrorResponse("invalid code"), nil
+		} else if err != nil {
+			return nil, err
+		}
+	} else if refreshToken, ok := data.GetOk("refresh_token"); ok {
+		tok = &oauth2.Token{RefreshToken: refreshToken.(string)}
```

Comment:
I think we also want to retrieve an initial access token here corresponding to the refresh token, otherwise we need to wait for the refresher to sync it. This would also let us validate the refresh token relative to the client information saved in our config. Assuming you have an exchange config builder `cb`, this could look something like:

```go
tok, err = cb.Build().TokenSource(ctx, &oauth2.Token{RefreshToken: refreshToken.(string)}).Token()
```

Reply:
Good idea, done.